# Causal Inference
# School of Information, University of Michigan 
## Week 1

### Resources:
 - Course Manual, which can be found in Coursera
 - [Intro to Pandas Data Structures](IntroPandasStructures.pdf)
 - [Bertrand and Mullainathan, 2004](BertrandMullainathan2004.pdf)

## Part 1

### Background

We will be using two datasets for this assignment. The first one (lecture1_survey.csv) is an observational dataset that comes from a labor market survey conducted in Chicago and Boston in 2001. The second one (lecture1_random_exper.csv) comes from a randomized experiment conducted by Marianne Bertrand and Sendhil Mullainathan, who sent fictitious resumes out to employers in response to job adverts in Boston and Chicago in 2001. The resumes differ in various attributes including the names of the applicants, and different resumes were randomly allocated to job openings. Some of the names are distinctly white sounding and some distinctly black sounding.


### Data

The data file “lecture_1_survey.csv” contains 4 variables for 10,593 individuals. Below are descriptions of each variable in the data:

- *black*: dummy variable for race; equal to 0 if white, 1 if black
- *yearsexp*: years of work experience
- *somecol_more*: 1 if college dropout or college degree or college degree and more, 0 otherwise
- *employed*: 1 is employed, 0 if unemployed

In [1]:
#Import statements. Run this cell.

import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.stats.api as sms
from scipy import stats
_TtestResult = type(stats.ttest_ind([1], [2]))
from mads.lib.path import assets
import warnings
warnings.simplefilter("ignore")

/usr/local/lib/python3.12/site-packages/scipy/stats/_stats_py.py:6312: RuntimeWarning: invalid value encountered in scalar divide
  svar = ((n1 - 1) * v1 + (n2 - 1) * v2) / df


In [2]:
#Run this cell to get the data for Part 1.
file = assets.find("lecture1_survey.csv")
data_survey = pd.read_csv(file)

#The line below lets you see the first five lines of the data_survey dataframe. Uncomment and run to view.
data_survey.head()

,yearsexp,somecol_more,col_more,employed,black
0,7,0,0,1,1
1,4,1,0,1,1
2,16,0,0,1,0
3,8,1,0,1,0
4,11,0,0,0,1


### Questions

We are interested in exploring whether there is racial discrimination against African-Americans in the labor market. We have two sets of data; observational (survey) and experimental that we can make use of for our investigation.

**Note:** You can refer to the manual for the methods we use in the assignment if you need to.

**Use the data_survey dataframe uploaded above to answer the questions in Part 1 unless otherwise specified.**

**1.** We will start by checking for covariate balance. Use the statsmodels.stats or scipy stats module to run a (two-sided) t-test to test the following null hypothesis: the mean of *somecol_more* is the same for African-Americans and Whites. Assign the results of the ttest to the variable `ttest1_1`. (1 pt)

**Note:** For this balance test, we could also use a nonparametric test (Chi-squared test for binary variables) that relies on fewer assumptions.

In [3]:
# split the education covariate by race, then compare group means with a two-sided t-test
# black = 1 represents African-Americans; black = 0 represents Whites
somecol_black = data_survey.loc[data_survey["black"] == 1, "somecol_more"]
somecol_white = data_survey.loc[data_survey["black"] == 0, "somecol_more"]

# the test asks whether the two group means are statistically distinguishable
ttest1_1 = stats.ttest_ind(somecol_black, somecol_white)

# display the key values needed for the written interpretation
ttest1_1

TtestResult(statistic=np.float64(-11.240955983443232), pvalue=np.float64(3.7514090837793756e-29), df=np.float64(10591.0))

In [4]:
# Hidden Tests, checking t-statistic and p-value of ttest1_1.

**2.** Based on the p-value, do you find evidence that this variable (somecol_more) differs between African-Americans and Whites at the 5% significance level? Explain. (i.e. report BOTH the p-value **AND** the decision rule based on p-value to determine if the difference is significant at the 5% level.) (1 pt)
 
**Note**: This question will be manually graded.

### STUDENT ANSWER

---

The p-value from the t-test is $3.7514 \times 10^{-29}$, which is much smaller than the 5% significance level, $ \alpha = 0.05$. 

The decision rule is: 
- If p-value < 0.05, reject the null hypothesis and conclude that the difference is statistically significant.
- If p-value $ \geq$ 0.05, fail to reject the null hypothesis and conclude that there is not enough evidence of a statistically significant difference.

Since $3.7514 \times 10^{-29} < 0.05$, we reject the null hypothesis. Thus, there is strong statistical evidence that `somecol_more` differes between African-American and White at the 5% significance level.

**3.** Use the statsmodel.stats or scipy stats module to run a (two-sided) t-test to test the following null hypothesis: the mean of years of experience (*yearsexp*) is the same for African-Americans and Whites. Assign the results of the ttest to the variable `ttest1_3`. (1 pt)

In [5]:
# split years of experience by race, then compare the average experience level
# this checks whether experience is balanced between African-Americans and Whites
yearsexp_black = data_survey.loc[data_survey["black"] == 1, "yearsexp"]
yearsexp_white = data_survey.loc[data_survey["black"] == 0, "yearsexp"]

# the two-sided t-test evaluates whether the difference in mean experience is plausibly zero
ttest1_3 = stats.ttest_ind(yearsexp_black, yearsexp_white)

# display the t-statistic and p-value for interpretation
ttest1_3

TtestResult(statistic=np.float64(-20.665913460336267), pvalue=np.float64(4.743084195168925e-93), df=np.float64(10591.0))

In [6]:
# Hidden Tests, checking the t-statistic and p-value of ttest1_3.

**4.** Based on the t-statistic, do you find evidence that this variable (yearsexp) differs between African-Americans and Whites at the 5% significance level? Explain. (i.e. report BOTH the t-statistic **AND** the decision rule based on t-statistic to determine if the difference is significant at the 5% level.) (1 pt) 

**Note**: This question will be manually graded

### STUDENT ANSWER

---

Based on the t-statistic, there is evidence that `yearsexp` differs between African-Americans and Whites at the 5% significance level.

The decision rule for a two-sided t-test at the 5% level is:
- Reject the null hypothesis if the t-statistic is less than the negative critical value or greater than the positive critical value.
- With a very large degrees of freedom value, df = 10591, the critical values are approx. -1.96 and 1.96.

Thsi rule is: `Reject H0 if t < -96 or t > 1.96`. Since the observed t-statistic is: `-20.6659 < -1.96`, we reject the null hypothesis that Afican-Americans and Whites have the same average `yearsexp`.

**5.** Discuss your results from 1 and 3 above. Why do we care about whether these variables look similar by race for the purpose of investigating racial discrimination in the labor market? (1 pt) 

**Note**: This question will be manually graded.

### STUDENT ANSWER

---

Questions 1 and 3 show that both `somecol_more` and `yearsexp` differ significantly between African-Americans and Whites. The p-value for `somecol_more` was far below 0.05, and the t-statistic for `yearsexp` was far outside the 5% critical values of about -1.96 and 1.96.

This matters because, when studying racial discrimination in the labor market, we want to compare groups that are otherwise similar on job-related characteristics. If variables like experience or other qualifications differ by race, then raw differences in labor market outcomes may reflect those factors in addition to possible discrimination. Since these variables are not balanced by race, we should control for them before making stronger causal claims about racial discrimination.

**6.** The main outcome variable is the indicator variable for employment status (*employed*). Use the statsmodel.stats or scipy stats module to run a (two-sided) t-test to test the null hypothesis that the means for the variable *employed* are the same for the two races.

In [7]:
# compare the main outcome, employment, across race groups
# this is a raw observational difference in employment rates, not automatically a causal effect
employed_black = data_survey.loc[data_survey["black"] == 1, "employed"]
employed_white = data_survey.loc[data_survey["black"] == 0, "employed"]

# the t-test checks whether mean employment differs between the two observed groups
ttest1_6 = stats.ttest_ind(employed_black, employed_white)

# display the result for the written decision rule
ttest1_6

TtestResult(statistic=np.float64(-7.1420030594715485), pvalue=np.float64(9.802071342176161e-13), df=np.float64(10591.0))

Based on the p-value above, do you find evidence that this variable differs significantly at 5% level between races? Explain. (i.e. report BOTH the p-value **AND** the decision rule based on p-value to determine if the difference is significant at the 5% level.) (1 pt) 

**Note**: This question will be manually graded.

### STUDENT ANSWER:

---

The p-value is: $9.8021 \times 10^{-13}$. At the 5% significance level, the decision rule is:

- If p-value $< 0.05%$, reject the null hypothesis.
- If p-value $ \geq 0.05$, fail to reject the null hypothesis.
  
Since $9.8021 \times 10^{-13} < 0.05$, we reject the null hypothesis. This means that there is a statisically significant evidence that the variable tested in #6 differs between African-Americans and Whites at the 5% level.

This t-statistic is -7.1420, which also falls far below the approximate lower critical value of -1.96, supporting the same conclusion. 

**7.** In light of your results, can we conclude from the observational (survey) data that there is racial discrimination against African-Americans in employment? Yes or no? Explain. (1 pt) 

**Note**: This question will be manually graded.

### STUDENT ANSWER

---

No, this is an observational study so we cannot causally conclude that there is racial discrimination against African Americans in employment.

Alrough results show that employement-related outcomes and background variables differ significantly by race, the data was not randomized, merely observational. This means that African Americans and Whites may differ on other important characteristics that could confound the actual relationshup between race and employment outcomes.

The survey merely shows an association between race and employment outcomes, but does not serve as difinitive proof of race caused the difference. An experiment that better controls for confounders could potentially provide a stronger causal claim of discrimination.

## Part 2
### Data

The data file “lecture1_random_exper.csv” contains 7 variables for 4,870 observations. Below are descriptions of each variable in the data:

* *id*: a de-identified identifier to represent unique observations (resumes)
* *male*: dummy variable for gender; equal to 0 if female, 1 if male
* *black*: dummy variable for race; equal to 0 if resume has a white sounding name and 1 if resume has a black sounding name
* *education*: 0 if not reported, 1 if high school dropout, 2 if high school graduate, 3 if college dropout, 4 if has college degree or more
* *computerskills*: 1 if resume mentions some computer skills, 0 otherwise
* *ofjobs*: number of jobs listed on the resume
* *yearsexp*: number of years of work experience on the resume
* *call*: 1 if applicant was called back, 0 otherwise

In [8]:
#Run this cell to get the data for Part 2.
file = assets.find("lecture1_random_exper.csv")
data_random = pd.read_csv(file)

#The line below lets you see the first five lines of the data_random. Uncomment and run to view.
data_random.head()

,education,ofjobs,yearsexp,computerskills,call,id,male,black
0,4,2,6,1,0,1,0,0
1,3,3,6,1,0,2,0,0
2,4,1,6,1,0,3,0,1
3,3,4,6,1,0,4,0,1
4,3,3,22,1,0,5,0,0


### Questions

In (Bertrand and Mullainathan 2004), researchers were interested to learn whether black sounding names obtain fewer callbacks for interviews than white sounding names.

**Use the data_random dataframe uploaded above to answer the questions in Part 2 unless otherwise specified.**

**1.** What is the treatment variable ($D_{i}$)? Assign the correct data_random column to the variable `d_i`. What is the outcome variable ($Y_{i}$)? Assign the correct data_random column to the variable `y_i`. (1 pt)

In [9]:
# the treatment is the race signal assigned to the resume name
# black = 1 means the resume has a Black-sounding name; black = 0 means a White-sounding name
d_i = data_random["black"]

# the outcome is whether the employer called the applicant back for an interview
y_i = data_random["call"]

In [10]:
# Hidden Tests, checking the columns assigned to d_i and y_i.

**2.** Remember the potential outcomes framework. By using the `loc` method, check for the treatment status of observation/individual id = 345. Based on what you see, what is the counterfactual outcome for observation/individual id =345, $Y_{345}^{black}$ or $Y_{345}^{white}$? (You can type “Y^b345” for $Y_{345}^{black}$ and “Y^w_345” for $Y_{345}^{white}$.) (1 pt)

**Note**: This question will be manually graded.

In [11]:
data_random.loc[data_random["id"] == 345, ["id", "black", "call"]]

,id,black,call
344,345,1,0


### STUDENT ANSWER

---

For observation 345, `black = 1`, meaning the resume was assigned a Black-sounding name. The observed outcome is therefore $Y_{345}^{black} = 0$, so this resume did not receive a callback. The counterfactual outcome is $Y_{345}^{white}$, which is what would have happened if the same resume had been assigned a White-sounding name. This counterfactual is not observed in the data.

**3.** Researchers designed the experiment such that the treatment and control groups are balanced on all other variables that can affect callback rates.

**3a.** Cross tabulate variables *male*, *computerskills*, *education*, and *ofjobs* with the variable *black*.

**Tip:** Use ``pd.crosstab()``. An example of a cross-tabulated table is shown below.


![Male variable cross-tabulated with black variable](crosstab_example.png)

In [12]:
# cross-tabs compare the distribution of each resume attribute across treatment groups
# because treatment was randomized, these distributions should look very similar by race
male_by_race = pd.crosstab(data_random["male"], data_random["black"])
computerskills_by_race = pd.crosstab(data_random["computerskills"], data_random["black"])
education_by_race = pd.crosstab(data_random["education"], data_random["black"])
ofjobs_by_race = pd.crosstab(data_random["ofjobs"], data_random["black"])

print("male by race")
display(male_by_race)

print("computer skills by race")
display(computerskills_by_race)

print("education by race")
display(education_by_race)

print("number of jobs by race")
display(ofjobs_by_race)

male by race


black,0,1
male,,
0,1860,1886
1,575,549


computer skills by race


black,0,1
computerskills,,
0,466,408
1,1969,2027


education by race


black,0,1
education,,
0,18,28
1,18,22
2,142,132
3,513,493
4,1744,1760


number of jobs by race


black,0,1
ofjobs,,
1,54,56
2,347,357
3,726,703
4,800,811
5,258,275
6,243,221
7,7,12


Do these variables look similar by race? (1 pt)

**Note**: This question will be manually graded.

### STUDENT ANSWER

---

The resume characteristics appears balanced by race, supporting the validity of the randomized experiment. For example, the counts for male applicants are very close across groups, as are the counts for computer skills, education levels, and number of previous jobs. This is important because these characteristics should not differ by much by race if race was randomly assigned. 

**3b.** Generate a table with the mean and standard deviation of the variable *yearsexp* for different values of black.

**Tip:** Use the ``groupby()`` and ``describe()`` methods.

In [13]:
# summarize years of experience by treatment group
# similar means and standard deviations provide another balance check
yearsexp_by_race = (
    data_random
    .groupby("black")["yearsexp"]
    .describe()[["mean", "std"]]
)

yearsexp_by_race

,mean,std
black,,
0,7.856263,5.079228
1,7.829569,5.010764


Does this variable look similar by race? (1 pt)

**Note**: This question will be manually graded.

### STUDENT ANSWER

---

The summary statistics show that the average value of this variable is very similar for White-sounding and Black-sounding resumes. This siggests that this characteristic is balanced across the two treatment groups, which supports the idea that race was randomly assigned. 

**4.** The outcome variable in the data set is the variable *call*, which indicates a callback for an interview. Let’s look at the means of this variable conditional on race.

**4a.** Use the statsmodel.stats or scipy stats module to run a (two-sided) t-test to test the following null hypothesis: the callback rates (means of variable *call*) are the same for resumes with black-sounding names as they are for resumes with white-sounding names. Assign the output to the variable `ttest2_4_1`. (1 pt)



In [14]:
# compare callback rates across the randomized race signal
# because race signal was randomly assigned, this difference is interpretable as causal evidence
call_black = data_random.loc[data_random["black"] == 1, "call"]
call_white = data_random.loc[data_random["black"] == 0, "call"]

# two-sided t-test of equal callback rates between Black-sounding and White-sounding names
ttest2_4_1 = stats.ttest_ind(call_black, call_white)

# display the result for interpretation
ttest2_4_1

TtestResult(statistic=np.float64(-4.114705266723095), pvalue=np.float64(3.94080251406952e-05), df=np.float64(4868.0))

In [15]:
# Hidden Tests, checking the t-statistic and p-value of ttest2_4_1.

**4b.** Based on the p-value, do you find evidence that there is discrimination in callback rates? That is, callback rates differ significantly at 5% level between resumes with white-sounding names and resumes with black-sounding names? Explain. (i.e. report BOTH the p-value **AND** the decision rule based on p-value to determine if the difference is significant at the 5% level.) (1 pt) 

**Note**: This question will be manually graded.

### STUDENT ANSWER:

---

At the 5% significance level, the decision rule based on the p-value is:
- If p-value $< 0.05$, reject the null hypothesis.
- If p-value $\geq 0.05$, fail to reject the null hypothesis.

Since p-value = $3.9408 \times 10^{-05} < 0.05$, we reject the null hypothesis. This means there is statistically significant evidence that the outcome tested in #4a differs between resumes assigned Black-sounding names and resumes assigned White-sounding names.

**5.** Let’s see if the results change when we compare callback rates conditional on gender.

**5a.** Use the statsmodel.stats or scipy stats module to carry out a (two-sided) t-test to test the hypothesis that callback rates are equal for females across races. Assign the result to the variable `ttest2_5_1`. (1 pt)



In [16]:
# restrict to resumes coded as female and compare callback rates across race signal
# this checks whether the racial callback gap appears within the female subgroup
female_black_calls = data_random.loc[
    (data_random["male"] == 0) & (data_random["black"] == 1),
    "call"
]

female_white_calls = data_random.loc[
    (data_random["male"] == 0) & (data_random["black"] == 0),
    "call"
]

# two-sided t-test of equal callback rates for female resumes across race
ttest2_5_1 = stats.ttest_ind(female_black_calls, female_white_calls)

ttest2_5_1

TtestResult(statistic=np.float64(-3.6369213964305627), pvalue=np.float64(0.00027963199420293636), df=np.float64(3744.0))

In [17]:
# Hidden Tests, checking the t-statistic and p-value of ttest2_5_1.

**5b.** Use the statsmodel.stats or scipy stats module to carry out a (two-sided) t-test to test the hypothesis that callback rates are equal for males across races. Assign the result to the variable ``ttest2_5_2``. (1 pt)

In [18]:
# restrict to resumes coded as male and compare callback rates across race signal
# this checks whether the racial callback gap appears within the male subgroup
male_black_calls = data_random.loc[
    (data_random["male"] == 1) & (data_random["black"] == 1),
    "call"
]

male_white_calls = data_random.loc[
    (data_random["male"] == 1) & (data_random["black"] == 0),
    "call"
]

# two-sided t-test of equal callback rates for male resumes across race
ttest2_5_2 = stats.ttest_ind(male_black_calls, male_white_calls)

ttest2_5_2

TtestResult(statistic=np.float64(-1.9501711134984252), pvalue=np.float64(0.05140448724722173), df=np.float64(1122.0))

In [19]:
# Hidden Tests, checking the t-statistic and p-value of ttest2_5_2.